pkg/virt-controller/watch/node_test.go

```diff
@@ -0,0 +1,359 @@
+
+		// Set up mock client
+		virtClient.EXPECT().VM(v1.NamespaceAll).Return(vmInterface).AnyTimes()
+		virtClient.EXPECT().VM(v1.NamespaceDefault).Return(vmInterface).AnyTimes()
+		kubeClient = fake.NewSimpleClientset()
+		virtClient.EXPECT().CoreV1().Return(kubeClient.CoreV1()).AnyTimes()
+
+		// Make sure that all unexpected calls to kubeClient will fail
+		kubeClient.Fake.PrependReactor("*", "*", func(action testing.Action) (handled bool, obj runtime.Object, err error) {
+			Expect(action).To(BeNil())
+			return true, nil, nil
+		})
+		syncCaches(stop)
+	})
+
+	addNode := func(node *k8sv1.Node) {
+		mockQueue.ExpectAdds(1)
+		nodeSource.Add(node)
+		mockQueue.Wait()
+	}
+
+	modifyNode := func(node *k8sv1.Node) {
+		mockQueue.ExpectAdds(1)
+		nodeSource.Modify(node)
+		mockQueue.Wait()
+	}
+
+	deleteNode := func(node *k8sv1.Node) {
+		mockQueue.ExpectAdds(1)
+		nodeSource.Delete(node)
+		mockQueue.Wait()
+	}
+
+	Context("pods and vms given", func() {
+		It("should only select stuck vms", func() {
+			node := NewHealthyNode("test")
+			finalVM := NewRunningVirtualMachine("finalVM", node)
+			finalVM.Status.Phase = virtv1.Succeeded
+			vmWithPod := NewRunningVirtualMachine("vmWithPod", node)
+			podForVM := NewHealthyPodForVirtualMachine("podForVM", vmWithPod)
+			vmWithPodInDifferentNamespace := NewRunningVirtualMachine("vmWithPodInDifferentNamespace", node)
+			podInDifferentNamespace := NewHealthyPodForVirtualMachine("podInDifferentnamespace", vmWithPodInDifferentNamespace)
+			podInDifferentNamespace.Namespace = "wrong"
+			vmWithoutPod := NewRunningVirtualMachine("vmWithoutPod", node)
+
+			vms := filterStuckVirtualMachinesWithoutPods([]*virtv1.VirtualMachine{
+				finalVM,
+				vmWithPod,
+				vmWithPodInDifferentNamespace,
+				vmWithoutPod,
+			}, []*k8sv1.Pod{
+				podForVM,
+				podInDifferentNamespace,
+			})
+
+			By("filtering out vms in final state")
+			Expect(vms).ToNot(ContainElement(finalVM))
+
+			By("filtering out vms which have a pod")
+			Expect(vms).ToNot(ContainElement(vmWithPod))
+
+			By("keeping vms which are running und have no pod in their namespace")
+			Expect(vms).To(ContainElement(vmWithoutPod))
+			Expect(vms).To(ContainElement(vmWithPodInDifferentNamespace))
+		})
+	})
+
+	Context("responsive virt-handler given", func() {
+		It("should do nothing", func() {
+			node := NewHealthyNode("testnode")
+
+			addNode(node)
+
+			controller.Execute()
+		})
+	})
+
+	Context("unresponsive virt-handler given", func() {
+		It("should set the node to unschedulable", func() {
+			node := NewHealthyNode("testnode")
+			node.Annotations[virtv1.VirtHandlerHeartbeat] = nowAsJSONWithOffset(-10 * time.Minute)
+
+			addNode(node)
+
+			kubeClient.Fake.PrependReactor("patch", "nodes", func(action testing.Action) (handled bool, obj runtime.Object, err error) {
+				return true, nil, nil
+			})
+
+			vmInterface.EXPECT().List(gomock.Any()).Return(&virtv1.VirtualMachineList{}, nil)
+
+			controller.Execute()
```

Comment:
Can we check that the `node` label equals to `kubevirt.io/schedulable: false`?

Reply:
Good catch. Done